flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/FlinkKafkaInternalProducerITCase.java

```diff
@@ -174,6 +177,51 @@ void testResetInnerTransactionIfFinalizingTransactionFailed(
         }
     }
 
+    @ParameterizedTest
+    @CsvSource({"true,true", "true,false", "false,true", "false,false"})
+    void testDoubleCommitAndAbort(boolean firstCommit, boolean secondCommit) {
+        final String topic = "test-double-commit-transaction-" + firstCommit + secondCommit;
+        final String transactionIdPrefix = "testDoubleCommitTransaction-";
+        final String transactionalId = transactionIdPrefix + "id";
+
+        KafkaCommittable committable;
+        try (FlinkKafkaInternalProducer<String, String> producer =
+                new FlinkKafkaInternalProducer<>(getProperties(), transactionalId)) {
+            producer.initTransactions();
+            producer.beginTransaction();
+            producer.send(new ProducerRecord<>(topic, "test-value"));
+            producer.flush();
+            committable = KafkaCommittable.of(producer);
+            if (firstCommit) {
+                producer.commitTransaction();
+            } else {
+                producer.abortTransaction();
+            }
+        }
+
+        try (FlinkKafkaInternalProducer<String, String> resumedProducer =
+                new FlinkKafkaInternalProducer<>(getProperties(), transactionalId)) {
+            resumedProducer.resumeTransaction(committable.getProducerId(), committable.getEpoch());
+            AbstractThrowableAssert<?, ? extends Throwable> secondOp =
+                    assertThatCode(
+                            () -> {
+                                if (secondCommit) {
+                                    resumedProducer.commitTransaction();
+                                } else {
+                                    resumedProducer.abortTransaction();
+                                }
+                            });
+            if (firstCommit == secondCommit) {
+                secondOp.doesNotThrowAnyException();
+            } else {
+                secondOp.isInstanceOf(InvalidTxnStateException.class);
+            }
+        }
+
+        assertNumTransactions(1, transactionIdPrefix);
+        assertThat(readRecords(topic).count()).isEqualTo(firstCommit ? 1 : 0);
```

Comment:
This is testing Producer, which wasn't actually changed in this PR, wasn't it?

Reply:
Yes, I was using the test to verify the expected behavior around kafka transactions (e.g. double commits are fine). It's not covering the changes but it's a test that I thought we already have. So I'd like to keep it.